rand_distr/src/zipf.rs

```diff
@@ -0,0 +1,328 @@
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! The Zeta and related distributions.
+
+use num_traits::Float;
+use crate::{Distribution, Standard};
+use rand::{Rng, distributions::OpenClosed01};
+use core::fmt;
+
+/// Samples integers according to the [zeta distribution].
+///
+/// The zeta distribution is a limit of the [`Zipf`] distribution. Sometimes it
+/// is called one of the following: discrete Pareto, Riemann-Zeta, Zipf, or
+/// Zipf–Estoup distribution.
+///
+/// It has the density function `f(k) = k^(-a) / C(a)` for `k >= 1`, where `a`
+/// is the parameter and `C(a)` is the Riemann zeta function.
+///
+/// # Example
+/// ```
+/// use rand::prelude::*;
+/// use rand_distr::Zeta;
+///
+/// let val: f64 = thread_rng().sample(Zeta::new(1.5).unwrap());
+/// println!("{}", val);
+/// ```
+///
+/// [zeta distribution]: https://en.wikipedia.org/wiki/Zeta_distribution
+#[derive(Clone, Copy, Debug)]
+pub struct Zeta<F>
+where F: Float, Standard: Distribution<F>, OpenClosed01: Distribution<F>
+{
+    a_minus_1: F,
+    b: F,
+}
+
+/// Error type returned from `Zeta::new`.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum ZetaError {
+    /// `a <= 1` or `nan`.
+    ATooSmall,
+}
+
+impl fmt::Display for ZetaError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(match self {
+            ZetaError::ATooSmall => "a <= 1 or is NaN in Zeta distribution",
+        })
+    }
+}
+
+#[cfg(feature = "std")]
+#[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
+impl std::error::Error for ZetaError {}
+
+impl<F> Zeta<F>
+where F: Float, Standard: Distribution<F>, OpenClosed01: Distribution<F>
+{
+    /// Construct a new `Zeta` distribution with given `a` parameter.
+    #[inline]
+    pub fn new(a: F) -> Result<Zeta<F>, ZetaError> {
+        if !(a > F::one()) {
+            return Err(ZetaError::ATooSmall);
+        }
+        let a_minus_1 = a - F::one();
+        let two = F::one() + F::one();
+        Ok(Zeta {
+            a_minus_1,
+            b: two.powf(a_minus_1),
+        })
+    }
+}
+
+impl<F> Distribution<F> for Zeta<F>
+where F: Float, Standard: Distribution<F>, OpenClosed01: Distribution<F>
+{
+    #[inline]
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> F {
+        // This is based on https://doi.org/10.1007/978-1-4613-8643-8.
+        loop {
+            let u = rng.sample(OpenClosed01);
+            let x = u.powf(-F::one() / self.a_minus_1).floor();
+            debug_assert!(x >= F::one());
+
+            let t = (F::one() + F::one() / x).powf(self.a_minus_1);
+
+            let v = rng.sample(Standard);
+            if v * x * (t - F::one()) / (self.b - F::one()) <= t / self.b {
```

Comment:
I think we can avoid the two divisions here by by multiplying the inequality with `b * (b - 1)`. This should be fine, because `b > 1`.

Reply:
Yes, indeed! :D